docs/RFCS/20191209_optimistic_ddl.md

```diff
@@ -0,0 +1,650 @@
+
+![architecture](../media/dm-architecture.png)
+
+One major feature of DM is [merging manually sharded tables](https://pingcap.com/docs/dev/reference/tools/data-migration/features/shard-merge/) into a single table.
+
+![shard-table-merge](../media/shard-table-merge.png)
+
+The design difficulty on merging sharded tables is handling schema change (DDLs). Currently, DM resolves this with pessimistic replication: if a DDL statement is executed on tbl01, a [barrier](https://en.wikipedia.org/wiki/Barrier_%28computer_science%29) (known as sharding DDL lock) would be set up to pause all binlogs from tbl01 (including this DDL). The same is applied on all other tables, until the same DDL statement is executed on all of tbl01 to tblNN, where the barrier would be unlocked.
+
+This approach ignores the actual action done by the DDL statement, e.g. even if executing ddl_1; ddl_2; is equivalent to ddl_3; ddl_4;, DM cannot recognize this and will not automatically unlock the barrier, and [require human intervention](https://pingcap.com/docs/dev/reference/tools/data-migration/features/manually-handling-sharding-ddl-locks/). This makes the whole feature feel risky and error-prone.
+
+Even strictly following DDL execution, if the user [takes a long time to execute all DDLs](https://docs.google.com/document/d/154r0RGBa5uTWsWV8Q1RmIezamcAdElcLQJ3z_7Vr4nk/edit#), it would create a very long backlog and thus cripples DM’s responsiveness.
+
+## Principles
+
+To prevent the drawbacks mentioned above to make DM more user-friendly, we want to [create a barrier-free algorithm](https://docs.google.com/document/d/19v05Cw6gWg3ccMOtP1EQfhpOirz4fF1mUNgChsmFlNw/edit#) (w.r.t. DMLs) for reconciling DDLs on sharded tables.
+
+Note: This work assumes each DM worker already [tracks the upstream schema](https://docs.google.com/document/d/1iP9yeylZfJaym_IWntn1Ge1t-_awfm23YNE-R-rP1s8/edit) independently from the downstream, so that column order is not a concern.
+
+### DDL awareness
+
+The reason DM chooses pessimistic replication (barrier) is because it is a general solution for all DDL kinds. DM did not need to special case. A barrier-free solution must require DM to understand the meaning of each DDL, however. As an example, blindly propagating a “drop column” DDL from tbl01 would cause error when replicating data from tbl02.
+
+```sql
+alter table tbl01 drop column col2;
+insert into tbl01 (col1, col3) values (11, 13);
+insert into tbl02 (col1, col2, col3) values (21, 22, 23);
+```
+
+In our case, we should at least recognize these DDL statements
+
+- alter table add column [first / after]
+
+- alter table drop column
+
+- create index
+
+- drop index
+
+- alter table modify column [first / after]
```

Comment:
How about adding ``` ` ``` wrapper for these lines?

Reply:
done in 2b44322